packages/autocomplete-core/src/onKeyDown.ts

```diff
@@ -22,39 +23,72 @@ export function onKeyDown<TItem extends BaseItem>({
   ...setters
 }: OnKeyDownOptions<TItem>): void {
   if (event.key === 'ArrowUp' || event.key === 'ArrowDown') {
-    // Default browser behavior changes the caret placement on ArrowUp and
-    // Arrow down.
-    event.preventDefault();
+    // eslint-disable-next-line no-inner-declarations
+    function triggerScrollIntoView() {
+      const nodeItem = props.environment.document.getElementById(
+        `${props.id}-item-${store.getState().activeItemId}`
+      );
 
-    store.dispatch(event.key, null);
+      if (nodeItem) {
+        if ((nodeItem as any).scrollIntoViewIfNeeded) {
+          (nodeItem as any).scrollIntoViewIfNeeded(false);
+        } else {
+          nodeItem.scrollIntoView(false);
+        }
+      }
+    }
 
-    const nodeItem = props.environment.document.getElementById(
-      `${props.id}-item-${store.getState().activeItemId}`
-    );
+    // eslint-disable-next-line no-inner-declarations
+    function triggerOnActive() {
+      const highlightedItem = getActiveItem(store.getState());
 
-    if (nodeItem) {
-      if ((nodeItem as any).scrollIntoViewIfNeeded) {
-        (nodeItem as any).scrollIntoViewIfNeeded(false);
-      } else {
-        nodeItem.scrollIntoView(false);
+      if (store.getState().activeItemId !== null && highlightedItem) {
+        const { item, itemInputValue, itemUrl, source } = highlightedItem;
+
+        source.onActive({
+          event,
+          item,
+          itemInputValue,
+          itemUrl,
+          refresh,
+          source,
+          state: store.getState(),
+          ...setters,
+        });
       }
     }
 
-    const highlightedItem = getActiveItem(store.getState());
-
-    if (store.getState().activeItemId !== null && highlightedItem) {
-      const { item, itemInputValue, itemUrl, source } = highlightedItem;
+    // Default browser behavior changes the caret placement on ArrowUp and
+    // Arrow down.
+    event.preventDefault();
 
-      source.onActive({
+    // When re-opening the panel, we need to split the logic to keep the actions
+    // synchronized as `onInput` returns a promise.
+    if (
+      store.getState().isOpen === false &&
+      (props.openOnFocus || Boolean(store.getState().query))
+    ) {
+      onInput({
         event,
-        item,
-        itemInputValue,
-        itemUrl,
+        props,
+        query: store.getState().query,
         refresh,
-        source,
-        state: store.getState(),
+        store,
         ...setters,
+      }).then(() => {
+        store.dispatch(event.key as ActionType, {
```

Comment:
I think it's because [onInput](https://github.com/algolia/autocomplete/pull/599/files/5287206cb998da64859b86c18f8829ec57be168f#diff-a33bd5a108ed1cc854e9a594b5ec2bef4313ea5d549ddd73473f841e35bed681R71-R78) is waiting for any type of event, so `event.key` is returned as a string (default type) instead of the previously defined `ActionType 'ArrowDown' | 'ArrowUp'`

Reply:
yes, it's because event is set as `any` here: https://github.com/algolia/autocomplete/blob/5287206cb998da64859b86c18f8829ec57be168f/packages/autocomplete-core/src/onInput.ts#L15